### DISCUSSION TITLE
Image Classification - Mengapa fit selalu stuck di tengah

### DISCUSSION THREAD

**Discussion by vetricko** [teamalgoritma/community#319] [Q&A] [answered]
sesuai dengan data di rpubs
code saya:

**data processing**
target_size <- c(100, 100)
batch_size <- 64

**data augmentation**
train_data_gen <- image_data_generator(rescale = 1/255, # Scaling pixel value
                                       horizontal_flip = T, # Flip image horizontally
                                       vertical_flip = T, # Flip image vertically 
                                       rotation_range = 45, # Rotate image from 0 to 45 degrees
                                       zoom_range = 0.25, # Zoom in or zoom out range
                                       validation_split = 0.2 # 20% data as validation data
                                       )

**splitting data:**
data train
train_image_array_gen <- flow_images_from_directory(directory = "train/", # Folder of the data
                                                    target_size = target_size, # target of the image dimension (100 x 100)  
                                                    color_mode = "rgb", # use RGB color
                                                    batch_size = batch_size , 
                                                    seed = 123,  # set random seed
                                                    subset = "training", # declare that this is for training data
                                                    generator = train_data_gen
                                                    )
data validation
val_image_array_gen <- flow_images_from_directory(directory = "train/",
                                                  target_size = target_size, 
                                                  color_mode = "rgb", 
                                                  batch_size = batch_size ,
                                                  seed = 123,
                                                  subset = "validation", # declare that this is the validation data
                                                  generator = train_data_gen
                                                  )

train_samples <- train_image_array_gen$n
valid_samples <- val_image_array_gen$n
output_n <- n_distinct(train_image_array_gen$classes)

**model:**
tensorflow::tf$random$set_seed(123)
model <- keras_model_sequential(name = "simple_model") %>% 
  layer_conv_2d(filters = 16,
                kernel_size = c(3,3),
                padding = "same",
                activation = "relu",
                input_shape = c(target_size, 3) 
                ) %>% 
  layer_max_pooling_2d(pool_size = c(2,2)) %>% 
  layer_flatten() %>% 
  layer_dense(units = 16,
              activation = "relu") %>% 
  layer_dense(units = output_n,
              activation = "softmax",
              name = "Output")

model %>% 
  compile(
    loss = "categorical_crossentropy",
    optimizer = optimizer_adam(learning_rate = 0.01),
    metrics = "accuracy"
  )

history <- model %>% 
  fit(
  train_image_array_gen,
  steps_per_epoch = as.integer(train_samples / batch_size), 
  epochs = 30, 
  validation_data = val_image_array_gen,
  validation_steps = as.integer(valid_samples / batch_size)
)

dengan target size 100x100
saya coba batch size 32: stuck di epoch 5/30 
<img width="677" alt="image" src="https://user-images.githubusercontent.com/65656955/191498276-0d228e5f-2531-483e-95a7-f088c53fcd90.png">

saya coba batch size 64: stuck di epoch 5/30
<img width="673" alt="image" src="https://user-images.githubusercontent.com/65656955/191500705-527d6f4f-fc79-47d6-b7db-ea140b1517e8.png">

saya coba batch size 166: stuck di epoch 6/10
<img width="682" alt="image" src="https://user-images.githubusercontent.com/65656955/191499212-0ce82065-f8fc-4465-bf33-74f7366315fa.png">

**Comment by NabiilahArdini** [comment]
Dear @vetricko , melihat warning message dari screenshoot Anda, sepertinya hal ini menyerupai bug yang didiskusikan di sini: https://github.com/tensorflow/tensorflow/issues/37876 
Namun untuk lama proses sepertinya berkaitan dengan memori device Anda.

**Reply by vetricko** [reply]
Makasih link yang telah diberikan @NabiilahArdini tetapi masalah nya masih sama..

Perbaikan yang sudah gua coba:
- Increase Target size: 256,256
- Increase Batch size: 200
- Verbose = 2
- Decreasing Learning rate: 0.001
- steps_per_execution=NULL (compile)

<img width="960" alt="image" src="https://user-images.githubusercontent.com/65656955/191654852-13232e0f-edbc-4e99-9260-416ab40e1f17.png">

ada yang bahas: "Turn off all callbacks leads to same on_train_batch_end time as with callbacks" tapi tidak tahu caranya.
ada juga yang bahas di hacking Callbacklist untuk remove munculnya warning, tapi setelah dicoba orang lain ternyata error jadi tidak gua lakukan.

sudah gua coba run 1 jam 3 menit saat ini, masih no progress

**Comment by vetricko** [comment, marked as answer]
Instead of using fit( ), use fit_generator( ). In fit_generator(), you don't pass the x and y directly, instead, they come from a generator. As it is written in [keras documentation](https://keras.io/models/sequential/), the generator is used when you want to avoid duplicate data when using multiprocessing. This is for practical purposes when you have a large dataset.
